interpreter/syntax/types.ml

```diff
@@ -41,7 +41,15 @@ and str_type =
   | DefArrayT of array_type
   | DefFuncT of func_type
 
-and sub_type = SubT of final * heap_type list * str_type
+and described_type =
+  | DescriptorT of heap_type * str_type
+  | NoDescriptorT of str_type
+
+and describing_type =
+  | DescribesT of heap_type * described_type
+  | NoDescribesT of described_type
+
+and sub_type = SubT of final * heap_type list * describing_type
```

Comment:
I would suggest a more direct syntax, which simplifies pattern matching a lot:
```suggestion
and desc_type = DescT of heap_type option * heap_type option * str_type

and sub_type = SubT of final * heap_type list * desc_type
```
(We could turn the tuple into a record to avoid confusing the two clauses, but I suspect that will be tedious to use.)